common/authorization/authorizer.go

```diff
@@ -103,3 +106,28 @@ func GetAuthProviderClient(privateKey string) (clientworker.AuthorizationProvide
 type FilteredRequestBody interface {
 	SerializeForLogging() (string, error)
 }
+
+func validatePermission(claims *JWTClaims, attributes *Attributes, data map[string]string) error {
+	groups := ""
+	switch attributes.Permission {
+	case PermissionRead:
+		groups = data[common.DomainDataKeyForReadGroups] + groupSeparator + data[common.DomainDataKeyForWriteGroups]
+	case PermissionWrite:
+		groups = data[common.DomainDataKeyForWriteGroups]
+	default:
+		return fmt.Errorf("token doesn't have permission for %v API", attributes.Permission)
+	}
+	// groups are separated by space
+	allowedGroups := strings.Split(groups, groupSeparator)    // groups that allowed by domain configuration(in domainData)
```

Comment:
if we are going to split the `groups` string here then I don't get why we use string concatenation to combine the groups. Better to keep a `var allowedGroups map[string]bool`

Reply:
changed code here. Added utility functions to deal with intersections a bit easier.